src/core/execute.c

```diff
@@ -116,7 +130,8 @@ void exec_context_tty_reset(const ExecContext *context, const ExecParameters *p)
 
         if (p && p->stdin_fd >= 0)
                 fd = p->stdin_fd;
-        else if (path) {
+        else if (context->tty_path || is_terminal_input(context->std_input) ||
+                        is_terminal_output(context->std_output) || is_terminal_output(context->std_error)) {
```

Comment:
would be nice to have a helper:

```c
static bool exec_context_uses_tty(const ExecContext *c) {
        …
````

that encapsulates this check.

but doesn't matter.